# SEMIR LiTS Reproduction — v4 oracle-first corrected

This version fixes the failure mode seen in `semir_lits_v3_run_summary.json`: the few-shot search selected an over-compressed graph with ~129 supernodes, oracle Dice ~0.03, and ~33% tumor deletion.

The core correction is: **never let a candidate win unless the graph can still recover the tumor**. Compression is a secondary objective after oracle Dice and tumor-retention are acceptable.


## 1. Setup

In [1]:

import numpy as np
import os, re, time, json, math, random
import fastloops

DATA_ROOT = "/scratch/ud3d4/acm_data/Data"

# Default HU window. The search below can be run under either [-50,250] or [0,200].
# Keep this global value synchronized with the selected search result before building graphs.
HU_MIN, HU_MAX = -50, 250

# These are overwritten by the oracle-first few-shot search in Section 4.
MERGE_DIST = 8       # ψ
CUT_DIST = 51        # α
DELETE_SMALL = 0     # β_min -- start conservative; tumor rind nodes must survive
DELETE_LARGE_FRAC = 1.00  # β_max = int(n_vox ** DELETE_LARGE_FRAC); 1.00 disables large-node deletion
VALUE_MIN = 0        # m_min -- disable intensity deletion until oracle is high
VALUE_MAX = 255      # m_max

# A strict 0.50 overlap threshold often yields almost no positive tumor supernodes
# when the graph is still imperfect. Sweep this after oracle is acceptable.
OVERLAP_THRESHOLD = 0.05

np.random.seed(42)
random.seed(42)

print("fastloops loaded")
print("Initial params will be overwritten by few-shot search:")
print(f"  HU=[{HU_MIN},{HU_MAX}]  ψ={MERGE_DIST}  α={CUT_DIST}")
print(f"  β_min={DELETE_SMALL}  β_max=n_vox**{DELETE_LARGE_FRAC}  m=[{VALUE_MIN},{VALUE_MAX}]")
print(f"  overlap_threshold={OVERLAP_THRESHOLD}")


fastloops loaded
Initial params will be overwritten by few-shot search:
  HU=[-50,250]  ψ=8  α=51
  β_min=0  β_max=n_vox**1.0  m=[0,255]
  overlap_threshold=0.05


## 2. Discover LiTS Volumes

In [2]:
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
import multiprocessing as mp

N_WORKERS = min(mp.cpu_count(), 32)  # use up to 32 cores
print(f"Available CPUs: {mp.cpu_count()}, using {N_WORKERS} workers")

def _check_tumor(vid):
    seg = np.load(os.path.join(DATA_ROOT, 'seg', f'segmentation-{vid}.npy'))
    return vid if (seg == 2).sum() > 0 else None

def discover_volumes():
    ct_dir = os.path.join(DATA_ROOT, 'ct')
    vids = []
    for f in sorted(os.listdir(ct_dir)):
        m = re.match(r'volume-(\d+)\.npy', f)
        if m:
            vid = int(m.group(1))
            if os.path.exists(os.path.join(DATA_ROOT, 'seg', f'segmentation-{vid}.npy')):
                vids.append(vid)
    # Parallel tumor check
    with ProcessPoolExecutor(max_workers=N_WORKERS) as pool:
        results = list(pool.map(_check_tumor, vids))
    return sorted([v for v in results if v is not None])

t0 = time.time()
all_vids = discover_volumes()
print(f"Found {len(all_vids)} LiTS volumes with tumor ({time.time()-t0:.1f}s)")

np.random.seed(42)
perm = np.random.permutation(len(all_vids))
n_train = int(0.7 * len(all_vids))
n_val = int(0.15 * len(all_vids))
train_ids = sorted([all_vids[i] for i in perm[:n_train]])
val_ids = sorted([all_vids[i] for i in perm[n_train:n_train + n_val]])
test_ids = sorted([all_vids[i] for i in perm[n_train + n_val:]])
ordered = train_ids + val_ids + test_ids
print(f"Split: {len(train_ids)} train / {len(val_ids)} val / {len(test_ids)} test")


Available CPUs: 128, using 32 workers


Found 118 LiTS volumes with tumor (0.3s)
Split: 82 train / 17 val / 19 test


## 3. Rust contraction/deletion helpers

The old notebook built all graphs once before the few-shot search. That is expensive and misleading. In this version the all-volume graph construction happens only **after** `Θopt` is selected.


In [3]:
def load_and_convert(vid, hu_min=None, hu_max=None):
    """Load raw CT and convert to uint8 with the requested HU window."""
    hu_min = HU_MIN if hu_min is None else hu_min
    hu_max = HU_MAX if hu_max is None else hu_max
    ct = np.load(os.path.join(DATA_ROOT, "ct", f"volume-{vid}.npy")).astype(np.float32)
    seg = np.load(os.path.join(DATA_ROOT, "seg", f"segmentation-{vid}.npy")).astype(np.int32)
    ct_u8 = np.clip(ct, hu_min, hu_max)
    ct_u8 = ((ct_u8 - hu_min) / (hu_max - hu_min) * 255).round().astype(np.uint8)
    ct_u8 = np.ascontiguousarray(ct_u8[..., np.newaxis])  # (D, H, W, 1)
    return ct, seg, ct_u8


def oracle_dice(labels_np, seg):
    """Quality ceiling of the graph minor.

    A perfect supernode classifier marks every surviving supernode that contains
    any tumor voxel as foreground. If this score is low, the coarsener has already
    destroyed the segmentation problem.
    """
    flat = labels_np.ravel()
    gt = (seg.ravel() == 2).astype(np.float64)
    gt_total = int(gt.sum())
    valid = flat >= 0
    if gt_total == 0 or not valid.any():
        return 0.0, 0, 100.0 if gt_total > 0 else 0.0
    max_id = int(flat[valid].max())
    tc = np.bincount(flat[valid], weights=gt[valid], minlength=max_id + 1)
    tumor_sids = np.where(tc > 0)[0]
    lut = np.zeros(max_id + 1, dtype=np.int32)
    lut[tumor_sids] = 1
    pred = np.where(valid, lut[flat], 0).reshape(labels_np.shape).astype(bool)
    gt_mask = seg == 2
    inter = int((pred & gt_mask).sum())
    dice = 2.0 * inter / (pred.sum() + gt_mask.sum() + 1e-8)
    deleted_tumor = int(gt[~valid].sum())
    deleted_pct = deleted_tumor / max(gt_total, 1) * 100.0
    return dice, len(tumor_sids), deleted_pct


def compute_intensity_std(labels_np, ct_u8):
    """Per-supernode intensity std in [0,1], indexed by supernode ID."""
    flat = labels_np.ravel()
    valid = flat >= 0
    if not valid.any():
        return np.array([], dtype=np.float32)
    max_id = int(flat[valid].max())
    vals = ct_u8[..., 0].ravel().astype(np.float64) / 255.0
    counts = np.bincount(flat[valid], minlength=max_id + 1).astype(np.float64)
    sums = np.bincount(flat[valid], weights=vals[valid], minlength=max_id + 1)
    sq_sums = np.bincount(flat[valid], weights=vals[valid] ** 2, minlength=max_id + 1)
    mean = sums / np.maximum(counts, 1.0)
    var = sq_sums / np.maximum(counts, 1.0) - mean ** 2
    return np.sqrt(np.maximum(var, 0.0)).astype(np.float32)


def boundary_mask_6conn(mask):
    """6-connected binary boundary without scipy dependency."""
    mask = mask.astype(bool)
    b = np.zeros_like(mask, dtype=bool)
    for axis in range(3):
        lo = [slice(None)] * 3
        hi = [slice(None)] * 3
        lo[axis] = slice(0, -1)
        hi[axis] = slice(1, None)
        diff = mask[tuple(lo)] != mask[tuple(hi)]
        b[tuple(lo)] |= diff
        b[tuple(hi)] |= diff
    return b & mask


def supernode_boundary_mask(labels_np):
    """Voxels adjacent to a different surviving supernode."""
    valid = labels_np >= 0
    b = np.zeros_like(labels_np, dtype=bool)
    for axis in range(3):
        lo = [slice(None)] * 3
        hi = [slice(None)] * 3
        lo[axis] = slice(0, -1)
        hi[axis] = slice(1, None)
        a = labels_np[tuple(lo)]
        c = labels_np[tuple(hi)]
        v = valid[tuple(lo)] & valid[tuple(hi)]
        diff = (a != c) & v
        b[tuple(lo)] |= diff
        b[tuple(hi)] |= diff
    return b & valid


def boundary_dice(labels_np, seg, target_label=2):
    """SEMIR few-shot objective: DSC(supernode boundaries, target GT boundary)."""
    gt_b = boundary_mask_6conn(seg == target_label)
    sn_b = supernode_boundary_mask(labels_np)
    denom = gt_b.sum() + sn_b.sum()
    if denom == 0:
        return 0.0
    return 2.0 * int((gt_b & sn_b).sum()) / (denom + 1e-8)


def run_minor(ct_u8, n_vox, params):
    """One wrapper around the Rust coarsener."""
    beta_max = int(n_vox ** float(params["beta_max_frac"]))
    return fastloops.merge_and_cut(
        ct_u8,
        merge_distance=int(params["psi"]),
        cut_distance=int(params["alpha"]),
        delete_small_node_max_size=int(params["beta_min"]),
        delete_large_node_min_size=beta_max,
        delete_value_min=int(params["m_min"]),
        delete_value_max=int(params["m_max"]),
        connectivity="faces",
    )

print("Helper functions defined.")


Helper functions defined.


In [4]:

# Optional baseline diagnostic only. Keep disabled for normal reproduction runs.
# The actual full graph build happens after the oracle-first few-shot search.
RUN_BASELINE_DIAGNOSTIC = False

if RUN_BASELINE_DIAGNOSTIC:
    def _build_minor_for_vid(vid):
        """Worker: build graph minor for one volume using the initial fallback params."""
        import fastloops
        ct_raw, seg, ct_u8 = load_and_convert(vid)
        n_vox = ct_raw.size
        t0 = time.time()
        raw_nf, raw_ei, raw_ef, raw_labels, raw_adj = fastloops.merge_and_cut(
            ct_u8, merge_distance=MERGE_DIST, cut_distance=CUT_DIST,
            delete_small_node_max_size=DELETE_SMALL,
            delete_large_node_min_size=int(n_vox ** DELETE_LARGE_FRAC),
            delete_value_min=VALUE_MIN, delete_value_max=VALUE_MAX,
            connectivity='faces',
        )
        dt = time.time() - t0
        labels_np = np.asarray(raw_labels)
        od, t_sn, del_pct = oracle_dice(labels_np, seg)
        return vid, raw_nf.shape[0], raw_ei.shape[1], od, del_pct, dt

    t0 = time.time()
    with ProcessPoolExecutor(max_workers=N_WORKERS) as pool:
        baseline_results = list(pool.map(_build_minor_for_vid, ordered))
    for vid, n_sn, n_edges, od, del_pct, dt in baseline_results:
        print(f"vol-{vid}: {n_sn:,} SN  {n_edges:,} edges  oracle={od:.4f}  del={del_pct:.1f}%  {dt:.1f}s")
    print(f"Baseline diagnostic done in {time.time()-t0:.1f}s")
else:
    print("Skipping pre-search all-volume graph build. This is intentional.")


Skipping pre-search all-volume graph build. This is intentional.


## 3b. Optional HU-window diagnostic

Run this only if the oracle-first search cannot find a candidate with acceptable oracle Dice. The goal is to compare whether `[0,200]` or `[-50,250]` gives cleaner tumor-preserving supernodes under conservative no-deletion settings.


In [5]:
RUN_HU_DIAGNOSTIC = False

if RUN_HU_DIAGNOSTIC:
    hu_configs = [
        (0, 200, "HU [0, 200] liver window"),
        (-50, 250, "HU [-50, 250] wider window"),
    ]
    test_vids = ordered[:min(20, len(ordered))]
    for hu_min, hu_max, label in hu_configs:
        oracles, sns, dels = [], [], []
        for vid in test_vids:
            ct_raw = np.load(os.path.join(DATA_ROOT, "ct", f"volume-{vid}.npy")).astype(np.float32)
            seg = np.load(os.path.join(DATA_ROOT, "seg", f"segmentation-{vid}.npy")).astype(np.int32)
            n_vox = ct_raw.size
            ct_u8 = np.clip(ct_raw, hu_min, hu_max)
            ct_u8 = ((ct_u8 - hu_min) / (hu_max - hu_min) * 255).round().astype(np.uint8)
            ct_u8 = np.ascontiguousarray(ct_u8[..., np.newaxis])
            nf, ei, ef, labels, adj = fastloops.merge_and_cut(
                ct_u8,
                merge_distance=8,
                cut_distance=64,
                delete_small_node_max_size=0,
                delete_large_node_min_size=n_vox,
                delete_value_min=0,
                delete_value_max=255,
                connectivity="faces",
            )
            od, t_sn, del_pct = oracle_dice(np.asarray(labels), seg)
            oracles.append(od); sns.append(nf.shape[0]); dels.append(del_pct)
        print(f"{label}:")
        print(f"  Oracle Dice: {np.mean(oracles):.4f} +/- {np.std(oracles):.4f}")
        print(f"  Supernodes:  {np.mean(sns):,.0f}")
        print(f"  Tumor deleted: {np.mean(dels):.2f}%")
        print()
else:
    print("Skipping optional HU diagnostic.")


Skipping optional HU diagnostic.


## 4. Oracle-first few-shot SEMIR parameter search

The previous run selected a candidate with too few nodes and near-zero oracle Dice. This corrected search uses hard/soft oracle-first scoring:

1. Reject or heavily demote candidates with low oracle Dice.
2. Reject candidates that delete tumor voxels.
3. Only then prefer paper-like compression.
4. Use a tumor-boundary band for boundary Dice so global background boundaries do not dominate.


In [6]:

# -------------------------
# Oracle-first few-shot parameter search — PARALLELIZED
# -------------------------
N_FEW = min(5, len(train_ids))
N_RANDOM = 250            # increase to 1000+ after the logic is validated
TOP_K_PRINT = 20
few_vids = train_ids[:N_FEW]
print(f"Few-shot volumes: {few_vids}")

# Search-space philosophy:
# - Keep β_min small so tumor rind nodes survive.
# - Disable intensity deletion first: m=[0,255]. Add intensity windows only after oracle is high.
# - Keep ψ moderate. ψ=42 over-compressed the graph in the previous run.
# - Set α as a multiple of ψ. α must be above ψ, but α=253 on uint8 effectively cuts almost nothing.
psi_values = [3, 4, 5, 6, 8, 10, 12, 15, 18, 22, 26]
beta_min_values = [0, 1, 2, 3, 5, 8, 13]
beta_max_frac_values = [0.96, 1.00]  # 1.00 effectively disables large-node deletion
alpha_multipliers = [2.5, 3.5, 5.0, 7.0, 10.0]
m_ranges = [(0, 255)]

# Deterministic grid candidates.
candidates = []
for psi in psi_values:
    for mult in alpha_multipliers:
        alpha = min(int(round(psi * mult)), 200)
        alpha = max(alpha, psi + 4)
        for beta_min in beta_min_values:
            for beta_max_frac in beta_max_frac_values:
                for m_min, m_max in m_ranges:
                    candidates.append(dict(
                        psi=int(psi), alpha=int(alpha), beta_min=int(beta_min),
                        beta_max_frac=float(beta_max_frac), m_min=int(m_min), m_max=int(m_max),
                    ))

# Add a small random tail, still constrained to tumor-preserving ranges.
rng = np.random.default_rng(42)
for _ in range(N_RANDOM):
    psi = int(rng.choice([3,4,5,6,7,8,9,10,12,14,16,18,20,22,24,26,30]))
    alpha = int(min(max(round(psi * rng.choice([2.0,2.5,3.0,3.5,4.5,6.0,8.0,10.0])), psi + 4), 220))
    beta_min = int(rng.choice([0,1,2,3,5,8,13,21]))
    beta_max_frac = float(rng.choice([0.94,0.96,0.98,1.00]))
    # Keep intensity deletion disabled in 80% of random candidates.
    if rng.random() < 0.80:
        m_min, m_max = 0, 255
    else:
        m_min = int(rng.choice([0,3,5,8,13]))
        m_max = int(rng.choice([242,250,255]))
    candidates.append(dict(psi=psi, alpha=alpha, beta_min=beta_min,
                           beta_max_frac=beta_max_frac, m_min=m_min, m_max=m_max))

# Deduplicate.
seen, uniq = set(), []
for c in candidates:
    key = tuple(c[k] for k in ['psi', 'alpha', 'beta_min', 'beta_max_frac', 'm_min', 'm_max'])
    if key not in seen:
        seen.add(key); uniq.append(c)
candidates = uniq
print(f"Evaluating {len(candidates)} candidates on {N_FEW} cases with {N_WORKERS} workers...")

# Pre-save few-shot data to /dev/shm for fast worker access.
FEW_DIR = '/dev/shm/semir_few_shot_oracle_first'
os.makedirs(FEW_DIR, exist_ok=True)
few_meta = []
for vid in few_vids:
    ct_raw, seg, ct_u8 = load_and_convert(vid)
    np.save(f'{FEW_DIR}/ct_u8_{vid}.npy', ct_u8)
    np.save(f'{FEW_DIR}/seg_{vid}.npy', seg)
    few_meta.append({'vid': vid, 'n_vox': ct_raw.size})


def _eval_candidate(args):
    """Worker: evaluate one candidate across few-shot volumes."""
    params, meta_list, few_dir = args
    import fastloops
    import numpy as _np

    def _dilate6(mask, iterations=3):
        mask = mask.astype(bool)
        out = mask.copy()
        for _ in range(iterations):
            m = out.copy()
            for ax in range(3):
                lo = [slice(None)] * 3; hi = [slice(None)] * 3
                lo[ax] = slice(0, -1); hi[ax] = slice(1, None)
                m[tuple(lo)] |= out[tuple(hi)]
                m[tuple(hi)] |= out[tuple(lo)]
            out = m
        return out

    def _boundary_dice_band(labels_np, seg):
        # Compare supernode boundaries only near the target tumor boundary.
        gt_mask = seg == 2
        valid = labels_np >= 0
        b_gt = _np.zeros_like(gt_mask, dtype=bool)
        b_sn = _np.zeros_like(labels_np, dtype=bool)
        for axis in range(3):
            lo = [slice(None)] * 3; hi = [slice(None)] * 3
            lo[axis] = slice(0, -1); hi[axis] = slice(1, None)
            d_gt = gt_mask[tuple(lo)] != gt_mask[tuple(hi)]
            b_gt[tuple(lo)] |= d_gt; b_gt[tuple(hi)] |= d_gt
            a = labels_np[tuple(lo)]; c = labels_np[tuple(hi)]
            v = valid[tuple(lo)] & valid[tuple(hi)]
            d_sn = (a != c) & v
            b_sn[tuple(lo)] |= d_sn; b_sn[tuple(hi)] |= d_sn
        band = _dilate6(b_gt, iterations=3) | gt_mask
        b_gt &= band
        b_sn &= band
        denom = b_gt.sum() + b_sn.sum()
        return 2.0 * int((b_gt & b_sn).sum()) / (denom + 1e-8) if denom > 0 else 0.0

    def _oracle_dice(labels_np, seg):
        flat = labels_np.ravel()
        gt = (seg.ravel() == 2).astype(_np.float64)
        gt_total = int(gt.sum())
        valid = flat >= 0
        if gt_total == 0 or not valid.any():
            return 0.0, 0, 100.0 if gt_total > 0 else 0.0
        max_id = int(flat[valid].max())
        tc = _np.bincount(flat[valid], weights=gt[valid], minlength=max_id + 1)
        tumor_sids = _np.where(tc > 0)[0]
        lut = _np.zeros(max_id + 1, dtype=_np.int32)
        lut[tumor_sids] = 1
        pred = _np.where(valid, lut[flat], 0).reshape(labels_np.shape).astype(bool)
        gt_mask = seg == 2
        inter = int((pred & gt_mask).sum())
        dice = 2.0 * inter / (pred.sum() + gt_mask.sum() + 1e-8)
        deleted_tumor = int(gt[~valid].sum())
        return dice, len(tumor_sids), deleted_tumor / max(gt_total, 1) * 100.0

    bds, ods, sns, dels, tumor_sns, edge_counts = [], [], [], [], [], []
    for m in meta_list:
        try:
            ct_u8 = _np.load(f"{few_dir}/ct_u8_{m['vid']}.npy")
            seg = _np.load(f"{few_dir}/seg_{m['vid']}.npy")
            n_vox = int(m['n_vox'])
            beta_max = int(n_vox ** float(params['beta_max_frac']))
            nf, ei, ef, labels, adj = fastloops.merge_and_cut(
                ct_u8,
                merge_distance=int(params['psi']),
                cut_distance=int(params['alpha']),
                delete_small_node_max_size=int(params['beta_min']),
                delete_large_node_min_size=beta_max,
                delete_value_min=int(params['m_min']),
                delete_value_max=int(params['m_max']),
                connectivity='faces',
            )
            labels_np = _np.asarray(labels)
            bd = _boundary_dice_band(labels_np, seg)
            od, t_sn, del_pct = _oracle_dice(labels_np, seg)
            bds.append(bd); ods.append(od); sns.append(nf.shape[0])
            dels.append(del_pct); tumor_sns.append(t_sn); edge_counts.append(ei.shape[1])
        except Exception:
            return None

    row = dict(params)
    row.update(
        boundary=float(_np.mean(bds)),
        oracle=float(_np.mean(ods)),
        sn=float(_np.mean(sns)),
        del_pct=float(_np.mean(dels)),
        tumor_sn=float(_np.mean(tumor_sns)),
        edges=float(_np.mean(edge_counts)),
    )

    # Oracle-first scoring. Candidates with low oracle cannot win.
    target_sn = 1500.0
    sn_term = -abs(_np.log((row['sn'] + 1.0) / target_sn))

    if row['oracle'] < 0.50:
        row['score'] = -1e6 + row['oracle']
    elif row['oracle'] < 0.75:
        row['score'] = -1e4 + 10.0 * row['oracle'] - 0.1 * row['del_pct']
    elif row['del_pct'] > 5.0:
        row['score'] = -1e3 + 10.0 * row['oracle'] - row['del_pct']
    else:
        row['score'] = (
            5.0 * row['oracle']
            + 1.0 * row['boundary']
            + 0.35 * sn_term
            - 0.05 * row['del_pct']
        )
    return row


t0 = time.time()
work_items = [(c, few_meta, FEW_DIR) for c in candidates]
with ProcessPoolExecutor(max_workers=N_WORKERS) as pool:
    raw_rows = list(pool.map(_eval_candidate, work_items))

rows = [r for r in raw_rows if r is not None]
elapsed = time.time() - t0
print(f"Evaluated {len(rows)}/{len(candidates)} candidates in {elapsed:.1f}s "
      f"({elapsed/max(len(candidates),1):.2f}s/candidate, {N_WORKERS} workers)")

if not rows:
    raise RuntimeError('No valid SEMIR candidates were evaluated.')

# Print top candidates by oracle first, then by final score.
print("\nTop candidates by ORACLE Dice:")
print(f"{'rank':>4s} {'score':>10s} {'bd':>7s} {'oracle':>7s} {'SN':>8s} {'del%':>6s} {'tumSN':>7s}  params")
for rank, r in enumerate(sorted(rows, key=lambda r: (r['oracle'], -r['del_pct'], -abs(np.log((r['sn']+1)/1500.0)), r['boundary']), reverse=True)[:TOP_K_PRINT], start=1):
    ptxt = f"ψ={r['psi']} α={r['alpha']} βmin={r['beta_min']} βmax=n^{r['beta_max_frac']:.2f} m=[{r['m_min']},{r['m_max']}]"
    print(f"{rank:4d} {r['score']:10.2f} {r['boundary']:7.4f} {r['oracle']:7.4f} {r['sn']:8.0f} {r['del_pct']:5.1f}% {r['tumor_sn']:7.1f}  {ptxt}")

rows_by_score = sorted(rows, key=lambda r: r['score'], reverse=True)
print("\nTop candidates by FINAL oracle-first score:")
for rank, r in enumerate(rows_by_score[:TOP_K_PRINT], start=1):
    ptxt = f"ψ={r['psi']} α={r['alpha']} βmin={r['beta_min']} βmax=n^{r['beta_max_frac']:.2f} m=[{r['m_min']},{r['m_max']}]"
    print(f"{rank:4d} score={r['score']:10.2f} bd={r['boundary']:.4f} oracle={r['oracle']:.4f} SN={r['sn']:.0f} del={r['del_pct']:.1f}% tumSN={r['tumor_sn']:.1f}  {ptxt}")

# Prefer candidates satisfying minimum viability. Fall back to best oracle with a loud warning.
viable = [r for r in rows if r['oracle'] >= 0.75 and r['del_pct'] <= 5.0 and r['tumor_sn'] >= 1]
if viable:
    BEST = sorted(viable, key=lambda r: r['score'], reverse=True)[0]
    print(f"\nSelected viable Θopt from {len(viable)} viable candidates.")
else:
    BEST = sorted(rows, key=lambda r: (r['oracle'], -r['del_pct'], r['boundary']), reverse=True)[0]
    print("\nWARNING: No candidate met oracle>=0.75 and tumor deletion<=5%.")
    print("Selected the best-oracle fallback so you can inspect diagnostics, but do NOT expect GINE training to reproduce the paper.")

MERGE_DIST = int(BEST['psi'])
CUT_DIST = int(BEST['alpha'])
DELETE_SMALL = int(BEST['beta_min'])
DELETE_LARGE_FRAC = float(BEST['beta_max_frac'])
VALUE_MIN = int(BEST['m_min'])
VALUE_MAX = int(BEST['m_max'])

print(f"\nSelected Θopt:")
print(json.dumps({k: BEST[k] for k in ['psi', 'alpha', 'beta_min', 'beta_max_frac',
                                        'm_min', 'm_max', 'boundary', 'oracle', 'sn', 'del_pct', 'tumor_sn']}, indent=2))

# Clean up shared memory.
import shutil
shutil.rmtree(FEW_DIR, ignore_errors=True)


Few-shot volumes: [0, 3, 4, 5, 6]
Evaluating 997 candidates on 5 cases with 32 workers...


Evaluated 997/997 candidates in 854.1s (0.86s/candidate, 32 workers)

Top candidates by ORACLE Dice:
rank      score      bd  oracle       SN   del%   tumSN  params
   1   -9993.02  0.3610  0.6982  1600645   0.1% 42305.6  ψ=3 α=8 βmin=0 βmax=n^0.96 m=[0,255]
   2   -9993.02  0.3610  0.6982  1600645   0.1% 42305.6  ψ=3 α=10 βmin=0 βmax=n^0.96 m=[0,255]
   3   -9993.02  0.3610  0.6982  1600645   0.1% 42305.6  ψ=3 α=15 βmin=0 βmax=n^0.96 m=[0,255]
   4   -9993.02  0.3610  0.6982  1600645   0.1% 42305.6  ψ=3 α=21 βmin=0 βmax=n^0.96 m=[0,255]
   5   -9993.02  0.3610  0.6982  1600645   0.1% 42305.6  ψ=3 α=30 βmin=0 βmax=n^0.96 m=[0,255]
   6   -9994.42  0.3616  0.5583  1335413   0.1% 34121.2  ψ=4 α=32 βmin=0 βmax=n^0.94 m=[8,242]
   7   -9994.61  0.3583  0.5389  1600646   0.0% 42305.8  ψ=3 α=8 βmin=0 βmax=n^1.00 m=[0,255]
   8   -9994.61  0.3583  0.5389  1600646   0.0% 42305.8  ψ=3 α=10 βmin=0 βmax=n^1.00 m=[0,255]
   9   -9994.61  0.3583  0.5389  1600646   0.0% 42305.8  ψ=3 α=15 βmin=0 βmax

## 5. Feature extraction and graph construction

This version uses:

- log volume
- log boundary/surface
- compactness
- elongation
- **principal-axis components** `axis_x, axis_y, axis_z`
- mean intensity
- true intensity std

That yields 9 node features for a single-channel CT volume. The paper lists dominant axis as one descriptor, but the actual axis is a 3D vector, so using its components is the least lossy implementation.


In [7]:
# Luke's feature extraction functions (from rust_crate.ipynb)

def _layout(C):
    return dict(area=0, s=[1, 2, 3],
                cov=[(4, 0, 0), (5, 1, 1), (6, 2, 2), (7, 0, 1), (8, 0, 2), (9, 1, 2)],
                chan0=10, boundary=10 + C + 6, D=3)


def node_invariants(node_feats, C=1, eps=1e-6):
    """Extract scale/orientation-invariant node features from Rust output."""
    f = node_feats.astype(np.float64)
    L = _layout(C)
    D = L["D"]
    N = f.shape[0]
    V = f[:, L["area"]]
    Vsafe = np.maximum(V, 1.0)
    mean_coord = np.stack([f[:, c] for c in L["s"]], axis=1) / Vsafe[:, None]

    cov = np.zeros((N, D, D))
    for col, i, j in L["cov"]:
        cij = f[:, col] / Vsafe - mean_coord[:, i] * mean_coord[:, j]
        cov[:, i, j] = cij; cov[:, j, i] = cij

    w = np.linalg.eigvalsh(cov)
    w = np.clip(w, 0.0, None)
    _, vec = np.linalg.eigh(cov)
    principal = vec[..., -1]
    trace = w.sum(axis=1)
    degenerate = trace < eps

    denom = w[:, 2] + eps
    shape = np.stack([(w[:, 2] - w[:, 1]) / denom,
                      (w[:, 1] - w[:, 0]) / denom,
                      w[:, 0] / denom], axis=1)
    shape[degenerate] = 0.0
    line_like = np.where(degenerate, 0.0, shape[:, 0])

    chan = f[:, L["chan0"]:L["chan0"] + C] / Vsafe[:, None] / 255.0
    compactness = f[:, L["boundary"]] / np.power(Vsafe, (D - 1.0) / D)
    elongation = np.where(w[:, 0] > eps, w[:, 2] / (w[:, 0] + eps), 1.0)
    elongation = np.clip(elongation, 1.0, 100.0)

    return dict(V=V, surface=f[:, L["boundary"]], centroid=mean_coord,
                eig=w, shape=shape, line_like=line_like, principal=principal,
                chan=chan, compactness=compactness, elongation=elongation,
                log_size=np.log(Vsafe))


def edge_invariants(node_feats, edge_index, edge_feats, C=1, eps=1e-6):
    """Extract scale-invariant edge features from Rust output."""
    inv = node_invariants(node_feats, C, eps)
    a = edge_index[0].astype(np.int64)
    b = edge_index[1].astype(np.int64)
    ef = edge_feats.astype(np.float64)
    blsafe = np.maximum(ef[:, 0], 1.0)

    size_contrast = np.abs(inv["V"][a] - inv["V"][b]) / (inv["V"][a] + inv["V"][b] + eps)
    bfrac_a = ef[:, 0] / (inv["surface"][a] + eps)
    bfrac_b = ef[:, 0] / (inv["surface"][b] + eps)
    mean_contrast = np.abs(inv["chan"][a] - inv["chan"][b])
    shape_dissim = np.abs(inv["shape"][a] - inv["shape"][b])
    axis_align = (np.abs(np.sum(inv["principal"][a] * inv["principal"][b], axis=1))
                  * np.minimum(inv["line_like"][a], inv["line_like"][b]))
    bcontrast = (ef[:, 1] / blsafe) / 255.0
    cut_frac = ef[:, 3] / blsafe

    cols = [size_contrast[:, None], bfrac_a[:, None], bfrac_b[:, None],
            mean_contrast if mean_contrast.ndim > 1 else mean_contrast[:, None],
            shape_dissim, axis_align[:, None], bcontrast[:, None], cut_frac[:, None]]
    return np.concatenate(cols, axis=1).astype(np.float32)

In [8]:
import torch
from torch_geometric.data import Data


def build_pyg_graph(raw_nf, raw_ei, raw_ef, labels_np, seg, ct_u8, C=1, overlap_threshold=OVERLAP_THRESHOLD):
    """Build a PyG graph from the Rust minor output.

    Labels are assigned by majority/overlap threshold on each supernode. The final
    metric is still lifted voxel Dice, so this threshold should be swept if oracle
    is good but training Dice is low.
    """
    n_sn = raw_nf.shape[0]
    n_edges = raw_ei.shape[1]

    inv = node_invariants(raw_nf, C)
    int_std = compute_intensity_std(labels_np, ct_u8)
    if len(int_std) < n_sn:
        int_std = np.pad(int_std, (0, n_sn - len(int_std)))
    int_std = int_std[:n_sn]

    principal = inv["principal"].astype(np.float32)
    # Eigenvectors have arbitrary sign. Canonicalize sign so the largest absolute
    # component is positive; this reduces random sign flips across cases.
    if len(principal):
        max_comp = np.argmax(np.abs(principal), axis=1)
        signs = np.sign(principal[np.arange(len(principal)), max_comp])
        signs[signs == 0] = 1
        principal = principal * signs[:, None]

    x = np.column_stack([
        np.log1p(inv["V"]),             # volume
        np.log1p(inv["surface"]),       # boundary/surface
        inv["compactness"],             # compactness
        inv["elongation"],              # elongation
        principal[:, 0],                 # dominant axis x
        principal[:, 1],                 # dominant axis y
        principal[:, 2],                 # dominant axis z
        inv["chan"][:, 0],              # mean intensity
        int_std,                         # intensity std
    ]).astype(np.float32)

    # Per-graph z-score normalization.
    for col in range(x.shape[1]):
        mu, sigma = float(x[:, col].mean()), float(x[:, col].std())
        if sigma > 1e-8:
            x[:, col] = (x[:, col] - mu) / sigma
        else:
            x[:, col] = 0.0

    if n_edges > 0:
        edge_attr = edge_invariants(raw_nf, raw_ei, raw_ef, C)
        ei_fwd = torch.tensor(raw_ei, dtype=torch.long)
        ei_rev = torch.stack([ei_fwd[1], ei_fwd[0]])
        edge_index = torch.cat([ei_fwd, ei_rev], dim=1)
        edge_attr_t = torch.tensor(np.concatenate([edge_attr, edge_attr]), dtype=torch.float32)
    else:
        edge_index = torch.zeros((2, 0), dtype=torch.long)
        edge_attr_t = torch.zeros((0, 10), dtype=torch.float32)

    flat = labels_np.ravel()
    valid = flat >= 0
    gt = (seg.ravel() == 2).astype(np.float64)
    max_id = int(flat[valid].max()) if valid.any() else -1
    y = np.zeros(n_sn, dtype=np.int64)
    if max_id >= 0:
        tumor_count = np.bincount(flat[valid], weights=gt[valid], minlength=max_id + 1)
        total_count = np.bincount(flat[valid], minlength=max_id + 1)
        overlap = tumor_count / np.maximum(total_count, 1)
        y[:min(n_sn, len(overlap))] = (overlap[:n_sn] >= overlap_threshold).astype(np.int64)

    return Data(
        x=torch.tensor(x, dtype=torch.float32),
        edge_index=edge_index,
        edge_attr=edge_attr_t,
        y=torch.tensor(y, dtype=torch.long),
    )


# Build graphs for all volumes using selected Θopt.
graphs = {}
raw_data = {}
oracles, sns, del_pcts = [], [], []

print(f"Building graphs with Θopt: ψ={MERGE_DIST}, α={CUT_DIST}, βmin={DELETE_SMALL}, βmax=n^{DELETE_LARGE_FRAC:.2f}, m=[{VALUE_MIN},{VALUE_MAX}]")
for vid in ordered:
    ct_raw, seg, ct_u8 = load_and_convert(vid)
    n_vox = ct_raw.size
    params = dict(psi=MERGE_DIST, alpha=CUT_DIST, beta_min=DELETE_SMALL,
                  beta_max_frac=DELETE_LARGE_FRAC, m_min=VALUE_MIN, m_max=VALUE_MAX)
    raw_nf, raw_ei, raw_ef, raw_labels, raw_adj = run_minor(ct_u8, n_vox, params)
    labels_np = np.asarray(raw_labels)

    data = build_pyg_graph(raw_nf, raw_ei, raw_ef, labels_np, seg, ct_u8, overlap_threshold=OVERLAP_THRESHOLD)
    graphs[vid] = data
    raw_data[vid] = {"labels": labels_np, "seg": seg, "n_sn": raw_nf.shape[0]}

    od, t_sn, del_pct = oracle_dice(labels_np, seg)
    oracles.append(od); sns.append(raw_nf.shape[0]); del_pcts.append(del_pct)
    n_tu = int((data.y == 1).sum())
    n_bg = int((data.y == 0).sum())
    print(f"vol-{vid}: {data.num_nodes:,} nodes ({n_tu} tumor, {n_bg:,} bg), "
          f"oracle={od:.4f}, del_tumor={del_pct:.1f}%, edges={data.num_edges:,}, edge_dim={data.edge_attr.shape[1] if data.edge_attr.numel() > 0 else 0}")

mean_oracle = float(np.mean(oracles))
mean_sn = float(np.mean(sns))
mean_del = float(np.mean(del_pcts))
print("\nGraph diagnostics:")
print(f"  Oracle Dice:    {mean_oracle:.4f} ± {np.std(oracles):.4f}")
print(f"  Mean supernodes:{mean_sn:,.0f} ± {np.std(sns):,.0f}")
print(f"  Tumor deleted:  {mean_del:.2f}%")
print("  Paper LiTS target: ~1,075 ± 297 supernodes")

if mean_oracle < 0.80:
    print("\nSTOP: Oracle is low. The coarsener is still the bottleneck; GINE training cannot reach 0.89 until this improves.")
    print("Recommended: increase N_RANDOM, expand low-psi candidates, test HU window, or relax deletion.")
    print("You may still run training for debugging, but treat the result as non-reproducible.")


/home/ud3d4/.conda/envs/llmft/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Building graphs with Θopt: ψ=3, α=8, βmin=0, βmax=n^0.96, m=[0,255]


vol-0: 259,432 nodes (400 tumor, 259,032 bg), oracle=0.0579, del_tumor=0.0%, edges=321,786, edge_dim=10


vol-3: 2,391,187 nodes (410 tumor, 2,390,777 bg), oracle=0.8803, del_tumor=0.0%, edges=2,919,376, edge_dim=10


vol-4: 2,285,046 nodes (205820 tumor, 2,079,226 bg), oracle=0.8547, del_tumor=0.3%, edges=3,025,084, edge_dim=10


vol-5: 1,381,323 nodes (95 tumor, 1,381,228 bg), oracle=0.7749, del_tumor=0.0%, edges=1,194,674, edge_dim=10


vol-6: 1,686,238 nodes (4775 tumor, 1,681,463 bg), oracle=0.9234, del_tumor=0.0%, edges=2,004,972, edge_dim=10


vol-7: 2,347,926 nodes (5135 tumor, 2,342,791 bg), oracle=0.9483, del_tumor=0.0%, edges=2,598,896, edge_dim=10


vol-8: 1,898,261 nodes (3162 tumor, 1,895,099 bg), oracle=0.9268, del_tumor=0.0%, edges=2,186,670, edge_dim=10


vol-9: 1,816,241 nodes (3416 tumor, 1,812,825 bg), oracle=0.9320, del_tumor=0.0%, edges=2,128,246, edge_dim=10


vol-10: 2,171,407 nodes (3751 tumor, 2,167,656 bg), oracle=0.9290, del_tumor=0.0%, edges=2,492,254, edge_dim=10


vol-11: 2,664,889 nodes (1770 tumor, 2,663,119 bg), oracle=0.9003, del_tumor=0.0%, edges=2,895,622, edge_dim=10


vol-12: 2,891,862 nodes (132 tumor, 2,891,730 bg), oracle=0.8184, del_tumor=0.0%, edges=3,406,772, edge_dim=10


vol-13: 1,221,964 nodes (2999 tumor, 1,218,965 bg), oracle=0.6917, del_tumor=0.0%, edges=1,522,152, edge_dim=10


vol-15: 1,975,386 nodes (151 tumor, 1,975,235 bg), oracle=0.7666, del_tumor=0.0%, edges=2,723,030, edge_dim=10


vol-16: 1,835,893 nodes (52458 tumor, 1,783,435 bg), oracle=0.9496, del_tumor=0.0%, edges=2,415,858, edge_dim=10


vol-17: 2,258,795 nodes (5351 tumor, 2,253,444 bg), oracle=0.9208, del_tumor=0.0%, edges=3,066,100, edge_dim=10


vol-18: 1,170,686 nodes (1224 tumor, 1,169,462 bg), oracle=0.7435, del_tumor=0.0%, edges=1,769,222, edge_dim=10


vol-19: 2,749,679 nodes (3634 tumor, 2,746,045 bg), oracle=0.9168, del_tumor=0.0%, edges=2,495,666, edge_dim=10


vol-22: 273,147 nodes (1096 tumor, 272,051 bg), oracle=0.8877, del_tumor=0.0%, edges=381,148, edge_dim=10


vol-24: 1,729,844 nodes (214 tumor, 1,729,630 bg), oracle=0.7764, del_tumor=0.0%, edges=2,138,422, edge_dim=10


vol-25: 1,630,433 nodes (76 tumor, 1,630,357 bg), oracle=0.8716, del_tumor=0.0%, edges=2,614,696, edge_dim=10


vol-26: 1,165,582 nodes (5663 tumor, 1,159,919 bg), oracle=0.9423, del_tumor=0.0%, edges=1,584,348, edge_dim=10


vol-27: 1,784,918 nodes (22276 tumor, 1,762,642 bg), oracle=0.9161, del_tumor=0.0%, edges=2,504,504, edge_dim=10


vol-28: 1,309,006 nodes (26032 tumor, 1,282,974 bg), oracle=0.7311, del_tumor=0.0%, edges=1,926,070, edge_dim=10


vol-30: 1,418,567 nodes (1948 tumor, 1,416,619 bg), oracle=0.9257, del_tumor=0.0%, edges=1,761,788, edge_dim=10


vol-31: 623,858 nodes (1007 tumor, 622,851 bg), oracle=0.0287, del_tumor=0.0%, edges=905,098, edge_dim=10


vol-35: 1,583,801 nodes (2740 tumor, 1,581,061 bg), oracle=0.8666, del_tumor=0.0%, edges=2,363,544, edge_dim=10


vol-36: 691,024 nodes (4489 tumor, 686,535 bg), oracle=0.9137, del_tumor=0.0%, edges=979,608, edge_dim=10


vol-37: 954,322 nodes (2427 tumor, 951,895 bg), oracle=0.8115, del_tumor=0.0%, edges=1,218,532, edge_dim=10


vol-39: 1,972,710 nodes (32808 tumor, 1,939,902 bg), oracle=0.9848, del_tumor=0.0%, edges=3,207,878, edge_dim=10


vol-42: 972,186 nodes (303 tumor, 971,883 bg), oracle=0.6596, del_tumor=0.0%, edges=853,282, edge_dim=10


vol-43: 1,644,178 nodes (1326 tumor, 1,642,852 bg), oracle=0.9337, del_tumor=0.0%, edges=1,461,236, edge_dim=10


vol-44: 1,357,468 nodes (22205 tumor, 1,335,263 bg), oracle=0.9434, del_tumor=0.2%, edges=1,646,344, edge_dim=10


vol-46: 334,342 nodes (6836 tumor, 327,506 bg), oracle=0.8073, del_tumor=0.1%, edges=390,068, edge_dim=10


vol-48: 472,765 nodes (3784 tumor, 468,981 bg), oracle=0.3455, del_tumor=0.0%, edges=681,450, edge_dim=10


vol-49: 452,458 nodes (1186 tumor, 451,272 bg), oracle=0.2246, del_tumor=0.0%, edges=621,844, edge_dim=10


vol-50: 394,990 nodes (467 tumor, 394,523 bg), oracle=0.8828, del_tumor=0.0%, edges=556,760, edge_dim=10


vol-51: 409,657 nodes (11433 tumor, 398,224 bg), oracle=0.5609, del_tumor=0.0%, edges=542,340, edge_dim=10


vol-52: 466,714 nodes (2230 tumor, 464,484 bg), oracle=0.7790, del_tumor=0.0%, edges=641,218, edge_dim=10


vol-54: 230,051 nodes (53 tumor, 229,998 bg), oracle=0.0195, del_tumor=0.0%, edges=316,814, edge_dim=10


vol-55: 1,077,081 nodes (446 tumor, 1,076,635 bg), oracle=0.9126, del_tumor=0.0%, edges=1,789,102, edge_dim=10


vol-58: 855,431 nodes (262 tumor, 855,169 bg), oracle=0.6580, del_tumor=0.0%, edges=1,120,146, edge_dim=10


vol-59: 1,117,683 nodes (124 tumor, 1,117,559 bg), oracle=0.8564, del_tumor=2.9%, edges=1,335,956, edge_dim=10


vol-60: 1,002,214 nodes (1878 tumor, 1,000,336 bg), oracle=0.9227, del_tumor=0.0%, edges=1,113,486, edge_dim=10


vol-61: 466,076 nodes (350 tumor, 465,726 bg), oracle=0.0452, del_tumor=0.0%, edges=772,406, edge_dim=10


vol-66: 469,325 nodes (353 tumor, 468,972 bg), oracle=0.1329, del_tumor=0.0%, edges=640,862, edge_dim=10


vol-67: 513,609 nodes (54 tumor, 513,555 bg), oracle=0.0037, del_tumor=0.0%, edges=848,584, edge_dim=10


vol-69: 1,042,252 nodes (543 tumor, 1,041,709 bg), oracle=0.9410, del_tumor=0.0%, edges=1,195,560, edge_dim=10


vol-70: 1,268,350 nodes (17076 tumor, 1,251,274 bg), oracle=0.8803, del_tumor=0.0%, edges=1,566,940, edge_dim=10


vol-71: 621,653 nodes (21084 tumor, 600,569 bg), oracle=0.9612, del_tumor=0.2%, edges=697,664, edge_dim=10


vol-72: 714,124 nodes (1939 tumor, 712,185 bg), oracle=0.2504, del_tumor=0.0%, edges=939,308, edge_dim=10


vol-73: 503,422 nodes (66 tumor, 503,356 bg), oracle=0.0784, del_tumor=0.0%, edges=668,132, edge_dim=10


vol-74: 378,433 nodes (4575 tumor, 373,858 bg), oracle=0.9484, del_tumor=0.0%, edges=522,626, edge_dim=10


vol-75: 378,238 nodes (357 tumor, 377,881 bg), oracle=0.1039, del_tumor=0.0%, edges=542,264, edge_dim=10


vol-77: 419,312 nodes (440 tumor, 418,872 bg), oracle=0.1567, del_tumor=0.0%, edges=545,010, edge_dim=10


vol-78: 510,010 nodes (1322 tumor, 508,688 bg), oracle=0.0489, del_tumor=0.0%, edges=835,498, edge_dim=10


vol-81: 1,160,105 nodes (810 tumor, 1,159,295 bg), oracle=0.9198, del_tumor=0.0%, edges=1,494,638, edge_dim=10


vol-82: 2,079,232 nodes (16162 tumor, 2,063,070 bg), oracle=0.9117, del_tumor=0.0%, edges=3,659,794, edge_dim=10


vol-83: 1,851,184 nodes (23 tumor, 1,851,161 bg), oracle=0.8788, del_tumor=0.0%, edges=1,987,714, edge_dim=10


vol-85: 4,655,126 nodes (3088 tumor, 4,652,038 bg), oracle=0.9656, del_tumor=0.0%, edges=3,115,090, edge_dim=10


vol-86: 4,454,066 nodes (998 tumor, 4,453,068 bg), oracle=0.8868, del_tumor=0.0%, edges=4,329,440, edge_dim=10


vol-90: 2,478,391 nodes (28866 tumor, 2,449,525 bg), oracle=0.9278, del_tumor=0.0%, edges=3,736,818, edge_dim=10


vol-92: 2,455,040 nodes (564 tumor, 2,454,476 bg), oracle=0.9074, del_tumor=0.0%, edges=2,741,716, edge_dim=10


vol-93: 2,653,869 nodes (44297 tumor, 2,609,572 bg), oracle=0.9315, del_tumor=0.0%, edges=3,862,480, edge_dim=10


vol-96: 4,729,909 nodes (7840 tumor, 4,722,069 bg), oracle=0.9265, del_tumor=0.0%, edges=4,460,660, edge_dim=10


vol-97: 3,528,039 nodes (127508 tumor, 3,400,531 bg), oracle=0.9685, del_tumor=0.0%, edges=3,289,694, edge_dim=10


vol-98: 3,025,639 nodes (101692 tumor, 2,923,947 bg), oracle=0.9787, del_tumor=0.0%, edges=2,876,258, edge_dim=10


vol-99: 2,346,423 nodes (2552 tumor, 2,343,871 bg), oracle=0.9511, del_tumor=0.5%, edges=2,038,850, edge_dim=10


vol-101: 4,982,090 nodes (67032 tumor, 4,915,058 bg), oracle=0.9729, del_tumor=0.0%, edges=4,441,452, edge_dim=10


vol-102: 5,329,541 nodes (8972 tumor, 5,320,569 bg), oracle=0.9766, del_tumor=0.0%, edges=3,882,956, edge_dim=10


vol-103: 2,263,741 nodes (13870 tumor, 2,249,871 bg), oracle=0.9782, del_tumor=0.0%, edges=1,733,222, edge_dim=10


vol-104: 1,475,503 nodes (28820 tumor, 1,446,683 bg), oracle=0.9670, del_tumor=0.1%, edges=2,463,632, edge_dim=10


vol-107: 2,053,009 nodes (807 tumor, 2,052,202 bg), oracle=0.8001, del_tumor=0.0%, edges=2,743,548, edge_dim=10


vol-111: 3,290,490 nodes (1277 tumor, 3,289,213 bg), oracle=0.8811, del_tumor=0.0%, edges=3,989,084, edge_dim=10


vol-113: 1,870,932 nodes (10649 tumor, 1,860,283 bg), oracle=0.8866, del_tumor=0.0%, edges=2,803,900, edge_dim=10


vol-117: 2,759,043 nodes (127978 tumor, 2,631,065 bg), oracle=0.9410, del_tumor=0.0%, edges=3,794,290, edge_dim=10


vol-120: 1,105,149 nodes (574 tumor, 1,104,575 bg), oracle=0.3723, del_tumor=0.0%, edges=1,835,890, edge_dim=10


vol-121: 1,238,195 nodes (344 tumor, 1,237,851 bg), oracle=0.8228, del_tumor=0.0%, edges=1,487,190, edge_dim=10


vol-122: 1,179,760 nodes (7655 tumor, 1,172,105 bg), oracle=0.5546, del_tumor=0.0%, edges=1,736,846, edge_dim=10


vol-124: 893,078 nodes (6852 tumor, 886,226 bg), oracle=0.3790, del_tumor=0.0%, edges=1,333,670, edge_dim=10


vol-125: 1,042,724 nodes (125 tumor, 1,042,599 bg), oracle=0.9258, del_tumor=0.0%, edges=1,492,224, edge_dim=10


vol-127: 2,694,527 nodes (58 tumor, 2,694,469 bg), oracle=0.8247, del_tumor=0.0%, edges=2,927,664, edge_dim=10


vol-129: 4,711,727 nodes (356522 tumor, 4,355,205 bg), oracle=0.9875, del_tumor=0.0%, edges=3,468,112, edge_dim=10


vol-1: 303,861 nodes (989 tumor, 302,872 bg), oracle=0.2619, del_tumor=0.1%, edges=380,942, edge_dim=10


vol-29: 1,004,705 nodes (1867 tumor, 1,002,838 bg), oracle=0.9345, del_tumor=0.0%, edges=1,203,754, edge_dim=10


vol-33: 1,230,505 nodes (73208 tumor, 1,157,297 bg), oracle=0.9593, del_tumor=0.0%, edges=1,162,470, edge_dim=10


vol-40: 1,445,336 nodes (20550 tumor, 1,424,786 bg), oracle=0.9088, del_tumor=0.0%, edges=1,590,676, edge_dim=10


vol-45: 439,774 nodes (713 tumor, 439,061 bg), oracle=0.9179, del_tumor=0.1%, edges=623,768, edge_dim=10


vol-53: 282,691 nodes (250 tumor, 282,441 bg), oracle=0.0733, del_tumor=0.0%, edges=382,450, edge_dim=10


vol-62: 896,087 nodes (486 tumor, 895,601 bg), oracle=0.7946, del_tumor=0.0%, edges=1,204,010, edge_dim=10


vol-63: 380,610 nodes (106 tumor, 380,504 bg), oracle=0.0644, del_tumor=0.0%, edges=495,208, edge_dim=10


vol-64: 1,249,443 nodes (25944 tumor, 1,223,499 bg), oracle=0.9768, del_tumor=0.1%, edges=1,500,928, edge_dim=10


vol-68: 648,136 nodes (714 tumor, 647,422 bg), oracle=0.8839, del_tumor=0.0%, edges=1,055,490, edge_dim=10


vol-80: 602,698 nodes (8374 tumor, 594,324 bg), oracle=0.5911, del_tumor=0.0%, edges=989,376, edge_dim=10


vol-84: 4,259,344 nodes (102149 tumor, 4,157,195 bg), oracle=0.9765, del_tumor=0.0%, edges=3,481,638, edge_dim=10


vol-108: 2,538,131 nodes (183131 tumor, 2,355,000 bg), oracle=0.9027, del_tumor=0.0%, edges=3,853,160, edge_dim=10


vol-110: 1,805,709 nodes (18008 tumor, 1,787,701 bg), oracle=0.9245, del_tumor=0.0%, edges=2,185,978, edge_dim=10


vol-116: 2,717,771 nodes (100266 tumor, 2,617,505 bg), oracle=0.9846, del_tumor=0.0%, edges=2,732,030, edge_dim=10


vol-123: 1,473,104 nodes (28959 tumor, 1,444,145 bg), oracle=0.9000, del_tumor=0.0%, edges=2,038,160, edge_dim=10


vol-128: 4,243,128 nodes (62347 tumor, 4,180,781 bg), oracle=0.9716, del_tumor=0.0%, edges=3,040,920, edge_dim=10


vol-2: 1,856,737 nodes (1833 tumor, 1,854,904 bg), oracle=0.9085, del_tumor=0.0%, edges=2,295,326, edge_dim=10


vol-14: 2,070,849 nodes (452 tumor, 2,070,397 bg), oracle=0.8960, del_tumor=0.0%, edges=2,649,638, edge_dim=10


vol-20: 2,475,344 nodes (332 tumor, 2,475,012 bg), oracle=0.8878, del_tumor=0.0%, edges=2,316,906, edge_dim=10


vol-21: 1,950,350 nodes (5910 tumor, 1,944,440 bg), oracle=0.9081, del_tumor=0.0%, edges=2,651,746, edge_dim=10


vol-23: 1,309,391 nodes (4681 tumor, 1,304,710 bg), oracle=0.9350, del_tumor=0.0%, edges=2,187,300, edge_dim=10


vol-56: 957,691 nodes (35518 tumor, 922,173 bg), oracle=0.9781, del_tumor=0.0%, edges=1,232,164, edge_dim=10


vol-57: 1,813,362 nodes (1043 tumor, 1,812,319 bg), oracle=0.8443, del_tumor=0.0%, edges=2,759,014, edge_dim=10


vol-65: 1,963,683 nodes (262 tumor, 1,963,421 bg), oracle=0.7686, del_tumor=0.0%, edges=2,887,628, edge_dim=10


vol-76: 685,071 nodes (17830 tumor, 667,241 bg), oracle=0.5381, del_tumor=0.0%, edges=1,038,696, edge_dim=10


vol-79: 576,787 nodes (1211 tumor, 575,576 bg), oracle=0.6742, del_tumor=0.0%, edges=758,920, edge_dim=10


vol-88: 1,893,279 nodes (31155 tumor, 1,862,124 bg), oracle=0.9553, del_tumor=0.0%, edges=2,809,830, edge_dim=10


vol-94: 3,826,227 nodes (11021 tumor, 3,815,206 bg), oracle=0.9318, del_tumor=0.0%, edges=4,085,940, edge_dim=10


vol-95: 2,602,131 nodes (347 tumor, 2,601,784 bg), oracle=0.7990, del_tumor=0.0%, edges=3,457,498, edge_dim=10


vol-100: 4,307,846 nodes (284342 tumor, 4,023,504 bg), oracle=0.9896, del_tumor=0.0%, edges=3,604,288, edge_dim=10


vol-109: 2,047,758 nodes (10822 tumor, 2,036,936 bg), oracle=0.9201, del_tumor=0.0%, edges=2,616,700, edge_dim=10


vol-112: 2,291,234 nodes (235 tumor, 2,290,999 bg), oracle=0.9305, del_tumor=0.0%, edges=3,166,870, edge_dim=10


vol-118: 1,515,539 nodes (63548 tumor, 1,451,991 bg), oracle=0.9694, del_tumor=0.0%, edges=1,634,858, edge_dim=10


vol-126: 1,576,092 nodes (882 tumor, 1,575,210 bg), oracle=0.9481, del_tumor=0.0%, edges=1,618,800, edge_dim=10


vol-130: 3,019,695 nodes (213149 tumor, 2,806,546 bg), oracle=0.9860, del_tumor=0.0%, edges=3,598,484, edge_dim=10

Graph diagnostics:
  Oracle Dice:    0.7701 ± 0.2840
  Mean supernodes:1,718,492 ± 1,168,702
  Tumor deleted:  0.04%
  Paper LiTS target: ~1,075 ± 297 supernodes

STOP: Oracle is low. The coarsener is still the bottleneck; GINE training cannot reach 0.89 until this improves.
Recommended: increase N_RANDOM, expand low-psi candidates, test HU window, or relax deletion.
You may still run training for debugging, but treat the result as non-reproducible.


## 6. GINE Training

Paper spec: 3-layer GINE, hidden 128, Adam lr=1e-3, patience 10.
Luke's additions: sqrt class weights (capped at 30), patience 30.

In [9]:
import torch.nn as nn
import torch.nn.functional as F
from torch_geometric.nn import GINEConv, BatchNorm


class GINE(nn.Module):
    """3-layer GINE — paper spec, with dynamic node/edge dimensions."""
    def __init__(self, node_dim, edge_dim, hidden=128):
        super().__init__()
        self.edge_proj = nn.Linear(edge_dim, hidden)

        def mlp(d_in):
            return nn.Sequential(
                nn.Linear(d_in, hidden), nn.BatchNorm1d(hidden), nn.ReLU(),
                nn.Linear(hidden, hidden),
            )

        self.conv1 = GINEConv(mlp(node_dim), edge_dim=hidden)
        self.bn1 = BatchNorm(hidden)
        self.conv2 = GINEConv(mlp(hidden), edge_dim=hidden)
        self.bn2 = BatchNorm(hidden)
        self.conv3 = GINEConv(mlp(hidden), edge_dim=hidden)
        self.bn3 = BatchNorm(hidden)
        self.head = nn.Linear(hidden, 2)

    def forward(self, data):
        x, ei, ea = data.x, data.edge_index, data.edge_attr
        if ea is not None and ea.numel() > 0:
            ea = self.edge_proj(ea)
        else:
            n = x.size(0)
            ei = torch.stack([torch.arange(n, device=x.device)] * 2)
            ea = torch.zeros(n, self.edge_proj.out_features, device=x.device)
        x = F.relu(self.bn1(self.conv1(x, ei, ea)))
        x = F.relu(self.bn2(self.conv2(x, ei, ea)))
        x = F.relu(self.bn3(self.conv3(x, ei, ea)))
        return self.head(x)


In [10]:

RUN_GINE_TRAINING = mean_oracle >= 0.75
if not RUN_GINE_TRAINING:
    print(f"Skipping GINE training because mean_oracle={mean_oracle:.4f} < 0.75.")
    print("Fix the graph minor first. Set RUN_GINE_TRAINING=True manually only for debugging.")
else:
    device = "cuda:0" if torch.cuda.is_available() else "cpu"
    print(f"Device: {device}")

    train_graphs = [graphs[v] for v in train_ids]
    val_graphs = [graphs[v] for v in val_ids]

    node_dim = train_graphs[0].x.shape[1]
    sample_ea = train_graphs[0].edge_attr
    edge_dim = sample_ea.shape[1] if sample_ea.numel() > 0 else 10
    print(f"Node dim: {node_dim}; edge dim: {edge_dim}")

    model = GINE(node_dim=node_dim, edge_dim=edge_dim).to(device)
    opt = torch.optim.Adam(model.parameters(), lr=1e-3)

    total_pos = sum(int((g.y == 1).sum()) for g in train_graphs)
    total_neg = sum(int((g.y == 0).sum()) for g in train_graphs)
    raw_ratio = total_neg / max(total_pos, 1)
    eff_ratio = min(np.sqrt(raw_ratio), 30.0)
    weight = torch.tensor([1.0, eff_ratio], dtype=torch.float32).to(device)
    criterion = nn.CrossEntropyLoss(weight=weight)
    print(f"Class weight: [1.0, {eff_ratio:.1f}] (raw ratio: {raw_ratio:.1f})")
    print(f"Tumor SN: {total_pos:,}; Background SN: {total_neg:,}")


    def lifted_voxel_dice_for_vids(model, vids, device):
        """Compute dataset-level lifted voxel Dice."""
        model.eval()
        tp = fp = fn = 0
        with torch.no_grad():
            for vid in vids:
                data = graphs[vid]
                rd = raw_data[vid]
                try:
                    logits = model(data.to(device))
                except RuntimeError:
                    torch.cuda.empty_cache()
                    continue
                preds = logits.argmax(dim=1).cpu().numpy()
                labels_np = rd["labels"]
                seg = rd["seg"]
                flat = labels_np.ravel()
                valid = flat >= 0
                pred_mask = np.zeros(labels_np.shape, dtype=bool)
                if valid.any():
                    max_id = int(flat[valid].max())
                    lut = np.zeros(max_id + 1, dtype=np.int8)
                    lut[:min(len(preds), max_id + 1)] = preds[:min(len(preds), max_id + 1)]
                    pred_mask = np.where(valid, lut[flat], 0).reshape(labels_np.shape).astype(bool)
                gt_mask = seg == 2
                inter = int((pred_mask & gt_mask).sum())
                tp += inter
                fp += int(pred_mask.sum()) - inter
                fn += int(gt_mask.sum()) - inter
        return 2 * tp / (2 * tp + fp + fn + 1e-8)


    PATIENCE = 30
    best_dice, best_state, wait = -1.0, None, 0
    history = {"train_loss": [], "val_voxel_dice": []}

    for epoch in range(1, 201):
        model.train()
        total_loss = 0.0
        for idx in np.random.permutation(len(train_graphs)):
            g = train_graphs[idx]
            if g.num_nodes > 100000:
                continue
            try:
                g = g.to(device)
                opt.zero_grad()
                loss = criterion(model(g), g.y)
                loss.backward()
                opt.step()
                total_loss += float(loss.item())
            except RuntimeError as e:
                if 'out of memory' in str(e):
                    torch.cuda.empty_cache()
                    continue
                raise
        mean_loss = total_loss / max(len(train_graphs), 1)
        history["train_loss"].append(mean_loss)

        val_dice = lifted_voxel_dice_for_vids(model, val_ids, device)
        history["val_voxel_dice"].append(val_dice)

        if val_dice > best_dice:
            best_dice = val_dice
            best_state = {k: v.detach().cpu().clone() for k, v in model.state_dict().items()}
            wait = 0
        else:
            wait += 1
            if wait >= PATIENCE:
                print(f"Early stop epoch {epoch}, best lifted val Dice={best_dice:.4f}")
                break

        if epoch % 10 == 0 or epoch == 1:
            print(f"Epoch {epoch:3d}  loss={mean_loss:.4f}  lifted_val_dice={val_dice:.4f}")

    if best_state is not None:
        model.load_state_dict(best_state)
    print(f"\nBest lifted voxel val Dice: {best_dice:.4f}")


Device: cuda:0
Node dim: 9; edge dim: 10


Class weight: [1.0, 9.7] (raw ratio: 93.7)
Tumor SN: 1,462,086; Background SN: 137,059,895


Epoch   1  loss=0.0000  lifted_val_dice=0.0446


Epoch  10  loss=0.0000  lifted_val_dice=0.0446


Epoch  20  loss=0.0000  lifted_val_dice=0.0446


Epoch  30  loss=0.0000  lifted_val_dice=0.0446


Early stop epoch 31, best lifted val Dice=0.0446

Best lifted voxel val Dice: 0.0446


## 7. Voxel-Level Evaluation

Lift supernode predictions to voxel grid via the label volume. Each voxel inherits its supernode's prediction.

In [11]:

if 'model' not in globals() or not RUN_GINE_TRAINING:
    print("No trained GINE model available. Saving graph-only diagnostics summary.")
    results = []
else:
    model.eval()
    model = model.to(device)
    results = []

    for vid in ordered:
        data = graphs[vid]
        rd = raw_data[vid]
        labels_np = rd["labels"]
        seg = rd["seg"]

        with torch.no_grad():
            try:
                preds = model(data.to(device)).argmax(dim=1).cpu().numpy()
            except RuntimeError:
                torch.cuda.empty_cache()
                preds = model.cpu()(data).argmax(dim=1).numpy()
                model = model.to(device)

        flat = labels_np.ravel()
        valid = flat >= 0
        pred_mask = np.zeros(labels_np.shape, dtype=bool)
        if valid.any():
            max_id = int(flat[valid].max())
            lut = np.zeros(max_id + 1, dtype=np.int8)
            lut[:min(len(preds), max_id + 1)] = preds[:min(len(preds), max_id + 1)]
            pred_mask = np.where(valid, lut[flat], 0).reshape(labels_np.shape).astype(bool)

        gt_mask = seg == 2
        inter = int((gt_mask & pred_mask).sum())
        dice = 2.0 * inter / (gt_mask.sum() + pred_mask.sum() + 1e-8)
        recall = inter / (gt_mask.sum() + 1e-8)
        precision = inter / (pred_mask.sum() + 1e-8) if pred_mask.sum() > 0 else 0.0

        split = "train" if vid in train_ids else ("val" if vid in val_ids else "test")
        results.append({"vid": vid, "split": split, "dice": dice, "recall": recall, "precision": precision})
        print(f"vol-{vid} [{split}]: Dice={dice:.4f}  Recall={recall:.4f}  Prec={precision:.4f}")

    print("\n" + "=" * 60)
    for split in ["train", "val", "test"]:
        scores = [r["dice"] for r in results if r["split"] == split]
        if scores:
            print(f"{split:>5s}: Dice = {np.mean(scores):.4f} ± {np.std(scores):.4f}  (n={len(scores)})")

print(f"\nOracle Dice:     {mean_oracle:.4f}")
print(f"Mean supernodes: {mean_sn:.0f}")
print(f"Tumor deleted:   {mean_del:.2f}%")
print("Paper target:    Dice 0.891 ± 0.007, ~1075 SN")

summary = {
    "theta": {"psi": MERGE_DIST, "alpha": CUT_DIST, "beta_min": DELETE_SMALL,
              "beta_max_frac": DELETE_LARGE_FRAC, "m_min": VALUE_MIN, "m_max": VALUE_MAX,
              "overlap_threshold": OVERLAP_THRESHOLD, "hu_min": HU_MIN, "hu_max": HU_MAX},
    "mean_oracle": mean_oracle,
    "mean_supernodes": mean_sn,
    "mean_tumor_deleted_pct": mean_del,
    "gine_training_ran": bool(RUN_GINE_TRAINING),
    "results": results,
}
with open("semir_lits_v4_run_summary.json", "w") as f:
    json.dump(summary, f, indent=2)
print("Saved semir_lits_v4_run_summary.json")


vol-0 [train]: Dice=0.0029  Recall=0.8935  Prec=0.0015


vol-3 [train]: Dice=0.0004  Recall=1.0000  Prec=0.0002


vol-4 [train]: Dice=0.1594  Recall=0.8635  Prec=0.0878


vol-5 [train]: Dice=0.0001  Recall=0.8649  Prec=0.0001


vol-6 [train]: Dice=0.0059  Recall=0.9445  Prec=0.0029


vol-7 [train]: Dice=0.0043  Recall=0.9370  Prec=0.0022


vol-8 [train]: Dice=0.0034  Recall=0.9221  Prec=0.0017


vol-9 [train]: Dice=0.0038  Recall=0.9386  Prec=0.0019


vol-10 [train]: Dice=0.0037  Recall=0.9394  Prec=0.0018


vol-11 [train]: Dice=0.0013  Recall=0.9467  Prec=0.0007


vol-12 [train]: Dice=0.0001  Recall=0.7710  Prec=0.0000
vol-13 [train]: Dice=0.0054  Recall=0.8730  Prec=0.0027


vol-15 [train]: Dice=0.0001  Recall=0.9424  Prec=0.0001


vol-16 [train]: Dice=0.0492  Recall=0.7820  Prec=0.0254


vol-17 [train]: Dice=0.0048  Recall=0.9097  Prec=0.0024


vol-18 [train]: Dice=0.0027  Recall=1.0000  Prec=0.0014


vol-19 [train]: Dice=0.0024  Recall=0.9641  Prec=0.0012
vol-22 [train]: Dice=0.0076  Recall=0.7628  Prec=0.0038


vol-24 [train]: Dice=0.0003  Recall=0.9791  Prec=0.0001


vol-25 [train]: Dice=0.0001  Recall=0.8605  Prec=0.0000


vol-26 [train]: Dice=0.0133  Recall=0.9784  Prec=0.0067


vol-27 [train]: Dice=0.0293  Recall=0.9052  Prec=0.0149
vol-28 [train]: Dice=0.0415  Recall=0.8890  Prec=0.0213


vol-30 [train]: Dice=0.0031  Recall=0.9973  Prec=0.0016
vol-31 [train]: Dice=0.0028  Recall=0.8611  Prec=0.0014


vol-35 [train]: Dice=0.0035  Recall=0.9772  Prec=0.0018
vol-36 [train]: Dice=0.0135  Recall=0.8134  Prec=0.0068


vol-37 [train]: Dice=0.0046  Recall=0.9443  Prec=0.0023


vol-39 [train]: Dice=0.0476  Recall=0.9990  Prec=0.0244
vol-42 [train]: Dice=0.0006  Recall=0.9634  Prec=0.0003


vol-43 [train]: Dice=0.0016  Recall=0.9989  Prec=0.0008
vol-44 [train]: Dice=0.0373  Recall=0.8894  Prec=0.0191


vol-46 [train]: Dice=0.0338  Recall=0.9282  Prec=0.0172
vol-48 [train]: Dice=0.0190  Recall=0.8694  Prec=0.0096
vol-49 [train]: Dice=0.0047  Recall=0.8631  Prec=0.0023


vol-50 [train]: Dice=0.0021  Recall=0.9115  Prec=0.0010
vol-51 [train]: Dice=0.0804  Recall=0.8970  Prec=0.0421
vol-52 [train]: Dice=0.0096  Recall=0.8562  Prec=0.0048
vol-54 [train]: Dice=0.0004  Recall=0.6331  Prec=0.0002


vol-55 [train]: Dice=0.0006  Recall=0.9197  Prec=0.0003
vol-58 [train]: Dice=0.0005  Recall=0.9921  Prec=0.0003


vol-59 [train]: Dice=0.0002  Recall=0.9714  Prec=0.0001
vol-60 [train]: Dice=0.0035  Recall=0.9358  Prec=0.0018


vol-61 [train]: Dice=0.0013  Recall=0.7858  Prec=0.0006
vol-66 [train]: Dice=0.0013  Recall=0.8728  Prec=0.0007
vol-67 [train]: Dice=0.0002  Recall=0.6197  Prec=0.0001


vol-69 [train]: Dice=0.0010  Recall=1.0000  Prec=0.0005
vol-70 [train]: Dice=0.0279  Recall=0.9642  Prec=0.0142


vol-71 [train]: Dice=0.0671  Recall=0.9633  Prec=0.0348
vol-72 [train]: Dice=0.0081  Recall=0.9710  Prec=0.0041
vol-73 [train]: Dice=0.0003  Recall=1.0000  Prec=0.0002


vol-74 [train]: Dice=0.0232  Recall=0.8380  Prec=0.0118
vol-75 [train]: Dice=0.0013  Recall=0.5934  Prec=0.0007
vol-77 [train]: Dice=0.0027  Recall=0.8714  Prec=0.0014


vol-78 [train]: Dice=0.0046  Recall=0.7373  Prec=0.0023
vol-81 [train]: Dice=0.0013  Recall=0.9726  Prec=0.0007


vol-82 [train]: Dice=0.0183  Recall=0.9614  Prec=0.0092


vol-83 [train]: Dice=0.0000  Recall=1.0000  Prec=0.0000


vol-85 [train]: Dice=0.0013  Recall=1.0000  Prec=0.0007


vol-86 [train]: Dice=0.0004  Recall=0.9714  Prec=0.0002


vol-90 [train]: Dice=0.0186  Recall=0.6155  Prec=0.0095


vol-92 [train]: Dice=0.0004  Recall=0.9848  Prec=0.0002


vol-93 [train]: Dice=0.0250  Recall=0.6459  Prec=0.0127


vol-96 [train]: Dice=0.0030  Recall=0.9960  Prec=0.0015


vol-97 [train]: Dice=0.0653  Recall=0.9888  Prec=0.0337


vol-98 [train]: Dice=0.0628  Recall=0.9782  Prec=0.0325


vol-99 [train]: Dice=0.0019  Recall=0.9903  Prec=0.0010


vol-101 [train]: Dice=0.0255  Recall=0.9971  Prec=0.0129


vol-102 [train]: Dice=0.0032  Recall=0.9999  Prec=0.0016


vol-103 [train]: Dice=0.0116  Recall=0.9889  Prec=0.0058


vol-104 [train]: Dice=0.0474  Recall=0.9455  Prec=0.0243


vol-107 [train]: Dice=0.0008  Recall=0.8431  Prec=0.0004


vol-111 [train]: Dice=0.0007  Recall=0.9647  Prec=0.0004


vol-113 [train]: Dice=0.0104  Recall=0.7734  Prec=0.0053


vol-117 [train]: Dice=0.0766  Recall=0.7243  Prec=0.0404
vol-120 [train]: Dice=0.0008  Recall=0.5676  Prec=0.0004


vol-121 [train]: Dice=0.0006  Recall=0.9842  Prec=0.0003
vol-122 [train]: Dice=0.0184  Recall=0.9431  Prec=0.0093


vol-124 [train]: Dice=0.0124  Recall=0.5448  Prec=0.0063
vol-125 [train]: Dice=0.0002  Recall=0.9854  Prec=0.0001


vol-127 [train]: Dice=0.0000  Recall=0.9875  Prec=0.0000


vol-129 [train]: Dice=0.1379  Recall=0.9994  Prec=0.0741
vol-1 [val]: Dice=0.0061  Recall=0.8457  Prec=0.0031
vol-29 [val]: Dice=0.0041  Recall=0.9464  Prec=0.0020


vol-33 [val]: Dice=0.1136  Recall=0.9916  Prec=0.0603
vol-40 [val]: Dice=0.0271  Recall=0.9847  Prec=0.0138


vol-45 [val]: Dice=0.0027  Recall=0.9570  Prec=0.0014
vol-53 [val]: Dice=0.0015  Recall=0.6163  Prec=0.0008
vol-62 [val]: Dice=0.0011  Recall=0.9939  Prec=0.0005
vol-63 [val]: Dice=0.0005  Recall=0.9247  Prec=0.0003


vol-64 [val]: Dice=0.0382  Recall=0.8527  Prec=0.0195
vol-68 [val]: Dice=0.0020  Recall=0.9345  Prec=0.0010
vol-80 [val]: Dice=0.0357  Recall=0.9289  Prec=0.0182


vol-84 [val]: Dice=0.0429  Recall=0.9986  Prec=0.0219


vol-108 [val]: Dice=0.1149  Recall=0.7121  Prec=0.0625


vol-110 [val]: Dice=0.0166  Recall=0.7698  Prec=0.0084


vol-116 [val]: Dice=0.0691  Recall=0.9665  Prec=0.0358
vol-123 [val]: Dice=0.0358  Recall=0.7779  Prec=0.0183


vol-128 [val]: Dice=0.0284  Recall=0.9984  Prec=0.0144


vol-2 [test]: Dice=0.0022  Recall=0.8826  Prec=0.0011


vol-14 [test]: Dice=0.0004  Recall=1.0000  Prec=0.0002


vol-20 [test]: Dice=0.0003  Recall=1.0000  Prec=0.0001


vol-21 [test]: Dice=0.0056  Recall=0.9399  Prec=0.0028


vol-23 [test]: Dice=0.0079  Recall=0.8738  Prec=0.0040
vol-56 [test]: Dice=0.0748  Recall=0.9158  Prec=0.0390


vol-57 [test]: Dice=0.0007  Recall=0.6636  Prec=0.0004


vol-65 [test]: Dice=0.0002  Recall=0.7169  Prec=0.0001
vol-76 [test]: Dice=0.0685  Recall=0.8714  Prec=0.0357
vol-79 [test]: Dice=0.0043  Recall=0.9147  Prec=0.0022


vol-88 [test]: Dice=0.0316  Recall=0.8391  Prec=0.0161


vol-94 [test]: Dice=0.0056  Recall=0.9856  Prec=0.0028


vol-95 [test]: Dice=0.0002  Recall=0.7077  Prec=0.0001


vol-100 [test]: Dice=0.1197  Recall=0.9981  Prec=0.0637


vol-109 [test]: Dice=0.0089  Recall=0.7895  Prec=0.0045


vol-112 [test]: Dice=0.0002  Recall=0.9909  Prec=0.0001


vol-118 [test]: Dice=0.0850  Recall=0.9651  Prec=0.0445


vol-126 [test]: Dice=0.0010  Recall=0.9938  Prec=0.0005


vol-130 [test]: Dice=0.1278  Recall=0.9565  Prec=0.0685

train: Dice = 0.0158 ± 0.0284  (n=82)
  val: Dice = 0.0318 ± 0.0355  (n=17)
 test: Dice = 0.0287 ± 0.0422  (n=19)

Oracle Dice:     0.7701
Mean supernodes: 1718492
Tumor deleted:   0.04%
Paper target:    Dice 0.891 ± 0.007, ~1075 SN
Saved semir_lits_v4_run_summary.json


## 8. How to interpret this run

The graph-only diagnostics decide whether training is meaningful.

- If `Oracle Dice < 0.75`, do **not** chase GINE training. Expand the search, lower `ψ`, lower/disable `β_min`, disable intensity deletion, or test the HU window.
- If `Tumor deleted > 5%`, deletion is too aggressive. Keep `β_min` small and `m=[0,255]`.
- If `Oracle Dice > 0.85` but `Mean supernodes` is huge, then start trading compression against oracle.
- If `Oracle Dice > 0.85` and `Mean supernodes` is roughly 1k–5k, then train GINE and sweep `OVERLAP_THRESHOLD` across `[0.01, 0.05, 0.10, 0.25, 0.50]`.

This notebook saves `semir_lits_v4_run_summary.json`.
